pkg/certificate/endpoint.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package certificate
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"storj.io/storj/pkg/certificate/authorization"
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+)
+
+// Endpoint implements pb.CertificatesServer.
+type Endpoint struct {
+	log             *zap.Logger
+	ca              *identity.FullCertificateAuthority
+	authorizationDB *authorization.DB
+	minDifficulty   uint16
+}
+
+// NewEndpoint creates a new certificate signing gRPC server.
+func NewEndpoint(log *zap.Logger, ca *identity.FullCertificateAuthority, authorizationDB *authorization.DB, minDifficulty uint16) *Endpoint {
+	return &Endpoint{
+		log:             log,
+		ca:              ca,
+		authorizationDB: authorizationDB,
+		minDifficulty:   minDifficulty,
+	}
+}
+
+// Sign signs the CA certificate of the remote peer's identity with the `certs.ca` certificate.
+// Returns a certificate chain consisting of the remote peer's CA followed by the CA chain.
+func (endpoint Endpoint) Sign(ctx context.Context, req *pb.SigningRequest) (_ *pb.SigningResponse, err error) {
+	defer mon.Task()(&ctx)(&err)
+	grpcPeer, ok := peer.FromContext(ctx)
+	if !ok {
+		msg := "error getting peer from context"
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, internalErr(msg)
+	}
+
+	peerIdent, err := identity.PeerIdentityFromPeer(grpcPeer)
+	if err != nil {
+		msg := "error getting peer identity"
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, internalErr(msg)
+	}
+
+	signedPeerCA, err := endpoint.ca.Sign(peerIdent.CA)
+	if err != nil {
+		msg := "error signing peer CA"
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, internalErr(msg)
+	}
+
+	signedChainBytes := [][]byte{signedPeerCA.Raw, endpoint.ca.Cert.Raw}
+	signedChainBytes = append(signedChainBytes, endpoint.ca.RawRestChain()...)
+	err = endpoint.authorizationDB.Claim(ctx, &authorization.ClaimOpts{
+		Req:           req,
+		Peer:          grpcPeer,
+		ChainBytes:    signedChainBytes,
+		MinDifficulty: endpoint.minDifficulty,
+	})
+	if err != nil {
+		msg := "error claiming authorization"
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, internalErr(msg)
+	}
+
+	difficulty, err := peerIdent.ID.Difficulty()
+	if err != nil {
+		msg := "error checking difficulty"
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, internalErr(msg)
+	}
+	token, err := authorization.ParseToken(req.AuthToken)
+	if err != nil {
+		msg := "error parsing auth token"
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, internalErr(msg)
+	}
+	tokenFormatter := authorization.Authorization{
+		Token: *token,
+	}
+	endpoint.log.Info("certificate successfully signed",
```

Comment:
```suggestion
	endpoint.log.Warn("certificate successfully signed",
```

Reply:
why warn instead of info?

Reply:
Because chatty logs make harder to find the important information (relevant errors, etc.).

In the end, it doesn't provide any information unless that you are specifically observing the service running, which usually is when you are trying to find some explanation to some bug/error or you're releasing a new release with important changes and you want to keep the service on eye for a while. In those situations, you change the configuration to run the service in `WARN` level, in normal conditions you don't.

Reply:
It's most seen from this point of view

* https://peter.bourgon.org/blog/2016/02/07/logging-v-instrumentation.html
* https://peter.bourgon.org/blog/2017/02/21/metrics-tracing-and-logging.html